File: src/engine/playerBoard.ts
/**
 * The player board: resource slots, city slots and the rewards they uncover
 * (rulebook p7, p17, p18).
 *
 * Cities sit in slots along the top of the board and are placed leftmost-first;
 * as they leave the board they uncover resource slots and the two "to won
 * ambitions" Power bonuses. Cities returning to the board cover them again.
 *
 * DATA-GAP: the rulebook states the mechanism and the +2/+5 bonus values but
 * not the slot layout or the per-slot raid costs — see docs/DATA-GAPS.md §6.
 */
import type { PlayerState, ResourceType } from './types';
import { RESOURCE_TYPES } from './types';

export const CITY_SLOTS = 5;
export const SHIPS = 15;
export const STARPORTS = 5;
export const AGENTS = 10;

/** Bonus Power for an outright ambition win, by uncovered slots (p18). */
export const BONUS_FIRST_ONE_SLOT = 2;
export const BONUS_FIRST_BOTH_SLOTS = 5;

/** What each city slot uncovers when its city is built. */
export type CityReward = { kind: 'resource' } | { kind: 'plusTwo' } | { kind: 'plusThree' };

export const PLAYER_BOARD = {
  citySlots: [
    { kind: 'resource' },
    { kind: 'resource' },
    { kind: 'plusTwo' },
    { kind: 'resource' },
    { kind: 'plusThree' },
  ] as CityReward[],
  /** Slots open before any city is built (p5 step O gives 2 starting resources). */
  baseResourceSlots: 3,
  /** Keys an attacker must spend to steal the resource in each slot (p16-p17). */
  raidCosts: [1, 1, 2, 2, 3, 3],
};

/** Total resource slots once every city is built. */
export const MAX_RESOURCE_SLOTS =
  PLAYER_BOARD.baseResourceSlots +
  PLAYER_BOARD.citySlots.filter((c) => c.kind === 'resource').length;

/** How many resource slots are currently open for a player. */
export function openResourceSlots(p: PlayerState): number {
  let n = PLAYER_BOARD.baseResourceSlots;
  for (let i = 0; i < p.citiesUsed && i < PLAYER_BOARD.citySlots.length; i++) {
    if (PLAYER_BOARD.citySlots[i].kind === 'resource') n++;
  }
  return Math.min(n, MAX_RESOURCE_SLOTS);
}

export function uncoveredBonuses(p: PlayerState): { plusTwo: boolean; plusThree: boolean } {
  let plusTwo = false;
  let plusThree = false;
  for (let i = 0; i < p.citiesUsed && i < PLAYER_BOARD.citySlots.length; i++) {
    if (PLAYER_BOARD.citySlots[i].kind === 'plusTwo') plusTwo = true;
    if (PLAYER_BOARD.citySlots[i].kind === 'plusThree') plusThree = true;
  }
  return { plusTwo, plusThree };
}

/** Raid cost of a held resource, by slot index. */
export function raidCost(slot: number): number {
  return PLAYER_BOARD.raidCosts[Math.min(slot, PLAYER_BOARD.raidCosts.length - 1)];
}

export function newPlayerState(): PlayerState {
  const outrage = {} as Record<ResourceType, boolean>;
  for (const r of RESOURCE_TYPES) outrage[r] = false;
  return {
    power: 0,
    resources: Array(MAX_RESOURCE_SLOTS).fill(null),
    outrage,
    guildCards: [],
    trophies: [],
    captives: [],
    agentsSupply: AGENTS,
    shipsSupply: SHIPS,
    starportsSupply: STARPORTS,
    citiesUsed: 0,
    hand: [],
  };
}

/**
 * Put a resource in the leftmost empty open slot. Returns false when the
 * player has no room — "you must discard resources you cannot hold" (p17).
 */
export function gainResource(p: PlayerState, r: ResourceType): boolean {
  const open = openResourceSlots(p);
  for (let i = 0; i < open; i++) {
    if (p.resources[i] === null) {
      p.resources[i] = r;
      return true;
    }
  }
  return false;
}

/** Discard resources sitting in slots that a returning city has covered. */
export function compactResources(p: PlayerState): void {
  const open = openResourceSlots(p);
  const held = p.resources.filter((r): r is ResourceType => r !== null).slice(0, open);
  p.resources = Array(MAX_RESOURCE_SLOTS).fill(null);
  held.forEach((r, i) => (p.resources[i] = r));
}

export function heldResources(p: PlayerState): ResourceType[] {
  return p.resources.filter((r): r is ResourceType => r !== null);
}

File: rust/crates/arcs-engine/src/player_board.rs
//! The player board: resource slots, city slots and the rewards they uncover
//! (rulebook p7, p17, p18), mirroring `src/engine/playerBoard.ts`.
//!
//! Cities sit in slots along the top of the board and are placed
//! leftmost-first; as they leave the board they uncover resource slots and
//! the two "to won ambitions" Power bonuses.

pub const CITY_SLOTS: usize = 5;
pub const SHIPS: u8 = 15;
pub const STARPORTS: u8 = 5;
pub const AGENTS: u8 = 10;

/// Bonus Power for an outright ambition win, by uncovered slots (p18).
pub const BONUS_FIRST_ONE_SLOT: u8 = 2;
pub const BONUS_FIRST_BOTH_SLOTS: u8 = 5;

/// What each city slot uncovers when its city is built.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum CityReward {
    Resource,
    PlusTwo,
    PlusThree,
}

/// The city-slot rewards, leftmost first.
pub const CITY_SLOT_REWARDS: [CityReward; CITY_SLOTS] = [
    CityReward::Resource,
    CityReward::Resource,
    CityReward::PlusTwo,
    CityReward::Resource,
    CityReward::PlusThree,
];

/// Slots open before any city is built (p5 step O gives 2 starting
/// resources).
pub const BASE_RESOURCE_SLOTS: usize = 3;

/// Keys an attacker must spend to steal the resource in each slot (p16-p17).
pub const RAID_COSTS: [u8; 6] = [1, 1, 2, 2, 3, 3];

/// Total resource slots once every city is built.
pub const MAX_RESOURCE_SLOTS: usize = {
    let mut n = BASE_RESOURCE_SLOTS;
    let mut i = 0;
    while i < CITY_SLOTS {
        if matches!(CITY_SLOT_REWARDS[i], CityReward::Resource) {
            n += 1;
        }
        i += 1;
    }
    n
};

/// How many resource slots are open for a player who has moved `cities_used`
/// cities off the board.
pub const fn open_resource_slots(cities_used: u8) -> usize {
    let mut n = BASE_RESOURCE_SLOTS;
    let mut i = 0;
    while i < cities_used as usize && i < CITY_SLOTS {
        if matches!(CITY_SLOT_REWARDS[i], CityReward::Resource) {
            n += 1;
        }
        i += 1;
    }
    if n > MAX_RESOURCE_SLOTS {
        MAX_RESOURCE_SLOTS
    } else {
        n
    }
}

/// The uncovered "to won ambitions" bonuses for `cities_used`.
pub const fn uncovered_bonuses(cities_used: u8) -> (bool, bool) {
    let mut plus_two = false;
    let mut plus_three = false;
    let mut i = 0;
    while i < cities_used as usize && i < CITY_SLOTS {
        match CITY_SLOT_REWARDS[i] {
            CityReward::PlusTwo => plus_two = true,
            CityReward::PlusThree => plus_three = true,
            CityReward::Resource => {}
        }
        i += 1;
    }
    (plus_two, plus_three)
}

/// Raid cost of a held resource, by slot index.
#[inline]
pub const fn raid_cost(slot: usize) -> u8 {
    let i = if slot < RAID_COSTS.len() {
        slot
    } else {
        RAID_COSTS.len() - 1
    };
    RAID_COSTS[i]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn max_resource_slots_is_six() {
        assert_eq!(MAX_RESOURCE_SLOTS, 6);
    }

    #[test]
    fn slots_open_as_cities_leave() {
        assert_eq!(open_resource_slots(0), 3);
        assert_eq!(open_resource_slots(1), 4);
        assert_eq!(open_resource_slots(2), 5);
        assert_eq!(open_resource_slots(3), 5); // slot 3 is the +2 bonus
        assert_eq!(open_resource_slots(4), 6);
        assert_eq!(open_resource_slots(5), 6); // slot 5 is the +3 bonus
        assert_eq!(open_resource_slots(9), 6);
    }

    #[test]
    fn bonuses_uncover_in_order() {
        assert_eq!(uncovered_bonuses(2), (false, false));
        assert_eq!(uncovered_bonuses(3), (true, false));
        assert_eq!(uncovered_bonuses(5), (true, true));
    }

    #[test]
    fn raid_costs_clamp() {
        assert_eq!(raid_cost(0), 1);
        assert_eq!(raid_cost(4), 3);
        assert_eq!(raid_cost(99), 3);
    }
}
